File: day8/day8.py
#!/usr/bin/env python3

import re

pat = re.compile(r"rect ([0-9]+)x([0-9]+)|rotate column x=([0-9]+) by ([0-9]+)|rotate row y=([0-9]+) by ([0-9]+)")

class Screen(object):
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.pixel = [[0 for x in range(width)] for y in range(height)]

    def rect(self, a, b):
        for y in range(b):
            for x in range(a):
                self.pixel[y][x] = 1

    def rotate_row(self, a, b):
        for i in range(b):
            right = self.pixel[a][-1]
            for x in range(self.width-1, 0, -1):
                self.pixel[a][x] = self.pixel[a][x-1]
            self.pixel[a][0] = right

    def rotate_col(self, a, b):
        for i in range(b):
            bottom = self.pixel[-1][a]
            for y in range(self.height-1, 0, -1):
                self.pixel[y][a] = self.pixel[y-1][a]
            self.pixel[0][a] = bottom

    def commands(self, txt):
        for row in txt.split("\n"):
            m = pat.search(row)
            if not m:
                continue
            g = m.groups()
            if g[0] and g[1]:
                self.rect(int(g[0]), int(g[1]))
            elif g[2] and g[3]:
                self.rotate_col(int(g[2]), int(g[3]))
            elif g[4] and g[5]:
                self.rotate_row(int(g[4]), int(g[5]))
            else:
                pass

    def lit(self):
        count = 0
        for row in self.pixel:
            count += sum(row)
        return count

    def __str__(self):
        lst = []
        for y, row in enumerate(self.pixel):
            lst.append("".join(v and "#" or " " for v in row))
        return "\n".join(lst)

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file = sys.stderr)
        sys.exit(1)

    s = Screen(50,6)
    s.commands(txt)
    print("Part1:", s.lit())
    print("Part2:")
    print(s)

File: day8/day8.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <regex>
#include <fmt/format.h>

using namespace std;

static const int WIDTH = 50;
static const int HEIGHT = 6;

static const regex rect("rect (\\d+)x(\\d+)");
static const regex rotrow("rotate row y=(\\d+) by (\\d+)");
static const regex rotcol("rotate column x=(\\d+) by (\\d+)");

struct Instruction
{
	enum { RECT, ROTR, ROTC } type;
	int arg1;
	int arg2;
};

istream& operator>>(istream& in, Instruction& i)
{
	string line;
	if (!getline(in, line))
	{
		return in;
	}
	smatch sm;
	if (regex_match(line, sm, rect))
	{
		i.type = Instruction::RECT;
	}
	else if (regex_match(line, sm, rotrow))
	{
		i.type = Instruction::ROTR;
	}
	else if (regex_match(line, sm, rotcol))
	{
		i.type = Instruction::ROTC;
	}
	else
	{
		in.setstate(ios::failbit);
		return in;
	}
	i.arg1 = stoi(sm[1]);
	i.arg2 = stoi(sm[2]);
	return in;
}

struct Screen
{
	char pixel[HEIGHT][WIDTH];

	Screen() : pixel{} {}

	void rect(size_t a, size_t b)
	{
		assert(a < WIDTH);
		assert(b < HEIGHT);
		for (size_t y = 0; y < b; y++)
		{
			for (size_t x = 0; x < a; x++)
			{
				pixel[y][x] = 1;
			}
		}
	}

	void rotate_row(size_t row, size_t b)
	{
		assert(row < HEIGHT);
		while (b-->0)
		{
			char end = pixel[row][WIDTH-1];
			for (size_t i = WIDTH-1; i > 0; i--)
			{
				pixel[row][i] = pixel[row][i-1];
			}
			pixel[row][0] = end;
		}
	}

	void rotate_col(size_t col, size_t b)
	{
		assert(col < WIDTH);
		while (b-->0)
		{
			char end = pixel[HEIGHT-1][col];
			for (size_t i = HEIGHT-1; i > 0; i--)
			{
				pixel[i][col] = pixel[i-1][col];
			}
			pixel[0][col] = end;
		}
	}

	void execute(const Instruction& inst)
	{
		switch(inst.type)
		{
		case Instruction::RECT: rect(inst.arg1, inst.arg2); break;
		case Instruction::ROTR: rotate_row(inst.arg1, inst.arg2); break;
		case Instruction::ROTC: rotate_col(inst.arg1, inst.arg2); break;
		}
	}

	size_t lit() const noexcept
	{
		return count(pixel[0], pixel[0]+WIDTH*HEIGHT, 1);
	}
};

ostream& operator<<(ostream& out, const Screen& s)
{
	for (size_t y = 0; y < HEIGHT; y++)
	{
		for (size_t x = 0; x < WIDTH; x++)
		{
			out << (s.pixel[y][x] ? '#' : ' ');
		}
		out << endl;
	}
	return out;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}
	Screen s;
	Instruction in;
	while (input>>in)
	{
		s.execute(in);
	}
	input.close();

	cout << fmt::format("Part1: {}", s.lit()) << endl
	     << "Part2:" << endl << s;
	return 0;
}
